src/python/WMCore/MicroService/Unified/MSRuleCleaner.py

```diff
@@ -50,27 +225,296 @@ def execute(self, reqStatus):
+            self.plineArchive.run(wflow)
+            self.wfCounters['archived'] += 1
+        except MSRuleCleanerArchival as ex:
+            msg = "%s: Archival Error: %s. "
+            msg += "\nWill retry again in the next cycle."
+            self.logger.error(msg, wflow['PlineMarkers'][-1], ex.myMessage)
+        except Exception as ex:
+            msg = "%s General error from pipeline. Error: \n%s. "
+            msg += "\nWill retry again in the next cycle."
+            self.logger.exception(msg, wflow['PlineMarkers'][-1], str(ex))
+        return wflow
+
+    def setPlineMarker(self, wflow, pName=None):
+        """
+        A function intended to mark which is the pipeline currently working
+        on the workflow. It is supposed to be called always as a first function
+        in the pipeline.
+        :param  wflow:   A MSRuleCleaner workflow representation
+        :param  pName:   The name of the functional pipeline
+        :return wflow:
+        """
+        # NOTE: The current functional pipeline MUST always be appended at the
+        #       end of the 'PlineMarkers' list
+
+        # First check if we can safely get rid of the default:
+        if len(wflow['PlineMarkers']) == 1 \
+           and wflow['PlineMarkers'][-1] is None \
+           and pName is not None:
+            wflow['PlineMarkers'].pop()
+
+        # Then push our current value into the markers list:
+        wflow['PlineMarkers'].append(pName)
+        return wflow
+
+    def _checkClean(self, wflow):
+        """
+        An auxiliary function used to only check the temporary cleanup status.
+        It basically takes the pipelines registered in 'PlineMarkers' that have
+        already worked on the workflow as a mask and applies this mask over
+        the set of flags in the 'CleanupStatus' field and then reduces the
+        result to a single bool value
+        """
+        # NOTE: This is one of the few functions taking a workflow as an argument
+        #       but returning a bool, since it is an auxiliary function and is not
+        #       supposed to be called as a standalone function in a pipeline.
+        # NOTE: `all([]) == True`, ergo all the 'rejected' && 'aborted-completed' workflows
+        #       are also counted as properly cleaned and can trigger archival later
+        cleanStatus = all([wflow['CleanupStatus'][key]
+                           for key in wflow['PlineMarkers']
+                           if key in wflow['CleanupStatus'].keys()])
+        return cleanStatus
+
+    def setClean(self, wflow):
+        """
+        A function to set the 'IsClean' flag based on the status from all the
+        pipelines which have worked on the workflow (and have put their markers
+        in the 'PlineMarkers' list)
+        :param  wflow:      A MSRuleCleaner workflow representation
+        :return wflow:
+        """
+        wflow['IsClean'] = self._checkClean(wflow)
+        return wflow
+
+    def archive(self, wflow):
```

Comment:
Aren't you missing the actual status transition here?

Reply:
This is to be done while working in a separate GH issue related to the archival only.

Reply:
Please make sure this is reflected in the PR description.